src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/NullableConverter.cs

```diff
@@ -1,73 +1,102 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Diagnostics;
-
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class NullableConverter<T> : JsonConverter<T?> where T : struct
     {
+        internal override ConverterStrategy ConverterStrategy { get; }
+        internal override Type? ElementType => typeof(T);
+        public override bool HandleNull => true;
+
         // It is possible to cache the underlying converter since this is an internal converter and
         // an instance is created only once for each JsonSerializerOptions instance.
-        private readonly JsonConverter<T> _converter;
+        private readonly JsonConverter<T> _elementConverter;
+
+        public NullableConverter(JsonConverter<T> elementConverter)
+        {
+            _elementConverter = elementConverter;
+            ConverterStrategy = elementConverter.ConverterStrategy;
+            IsInternalConverterForNumberType = elementConverter.IsInternalConverterForNumberType;
+            // temporary workaround for JsonConverter base constructor needing to access
+            // ConverterStrategy when calculating `CanUseDirectReadOrWrite`.
+            CanUseDirectReadOrWrite = elementConverter.ConverterStrategy == ConverterStrategy.Value;
+        }
+
+        internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out T? value)
+        {
+            if (!state.IsContinuation && reader.TokenType == JsonTokenType.Null)
+            {
+                value = null;
+                return true;
+            }
 
-        public NullableConverter(JsonConverter<T> converter)
+            state.Current.JsonPropertyInfo = state.Current.JsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
+            if (_elementConverter.TryRead(ref reader, typeof(T), options, ref state, out T element))
+            {
+                value = element;
+                return true;
+            }
+
+            value = null;
+            return false;
```

Comment:
could you just pass in `value` directly here rather than `out T element`? wouldn't it return null when the read returns false? this could then be simplified to:

```suggestion
            return _elementConverter.TryRead(ref reader, typeof(T), options, ref state, out value);
```